### PR TITLE
Fix Zc* support for writable misa

### LINKED ISSUE #1172: Alignment  when 'C'  in misa CSR is cleared by software
Whether jump to 0x10082 triggers misaligned exception when  'C' in misa CSR is disabled by software? 
The answer is yes before Zc* is supported. 
However,  Zca is implied by 'C', but Zca can work alone without 'C' enabled.  So this becomes a new problem:
Whether Zca is also disabled when 'C' is disabled?

### PR DESCRIPTION
**Pull request by liweiwei90**
Fix https://github.com/riscv-software-src/riscv-isa-sim/issues/1172
- Disable Zca/Zcf/Zcd when misa.C is cleared if Zcb/Zcmt/Zcmp is not supported
- Enable Zca/Zcf/Zcd when misa.C is enabled
- Make Zcf/Zcd only work when misa.F/D enabled

### CODE REVIEW

File: riscv/processor.h [hunk 1]
```diff
@@ -267,7 +276,7 @@ class processor_t : public abstract_device_t
   const char* get_symbol(uint64_t addr);
 
 private:
-  const isa_parser_t * const isa;
+  isa_parser_t * const isa;
```
Comment (aswaterman) [inline riscv/processor.h]: This doesn't seem right.  I think the parsed ISA string should be immutable; it is the processor whose state changes when `misa` changes.

File: riscv/insns/c_fld.h [hunk 2]
```diff
@@ -1,3 +1,4 @@
 require_extension(EXT_ZCD);
+require_extension('D');
```
Comment (aswaterman) [inline riscv/insns/c_fld.h]: The load/store changes should be a separate commit.

File: riscv/processor.h (hunk 1 above)
Comment (liweiwei90) [reply riscv/processor.h]: OK. Maybe we can maintain a disable list in processor.

File: riscv/insns/c_fld.h (hunk 2 above)
Comment (liweiwei90) [reply riscv/insns/c_fld.h]: We can not simply enable Zcd when we set misa.D if we disable it when we clear misa.D in current implementation. There are two different cases: D_Zca_Zcd and D_Zca. So I do nothing for Zcd when misa.D changes and add these checks.

File: riscv/csrs.cc [hunk 3]
```diff
@@ -620,20 +620,49 @@ reg_t misa_csr_t::dependency(const reg_t val, const char feature, const char dep
...
+      // the write is ignored if increasing IALIGN would misalign the PC
+      if (state->pc & 2)
+        return false;
+      proc->disable_extension(EXT_ZCA);
+      proc->disable_extension(EXT_ZCF);
+      proc->disable_extension(EXT_ZCD);
+    }
+  } else if (new_c && !prev_c) {
+    proc->enable_extension(EXT_ZCA);
+  }
```
Comment (scottj97) [inline riscv/csrs.cc]: Does it make sense to disable 3 extensions but only re-enable one of the three?

File: riscv/csrs.cc [hunk 4]
```diff
@@ -620,20 +620,49 @@ reg_t misa_csr_t::dependency(const reg_t val, const char feature, const char dep
...
+
+  const bool prev_f = old_misa & (1L << ('F' - 'A'));
+  const bool new_f = new_misa & (1L << ('F' - 'A'));
+  if (!new_f && prev_f) {
+    proc->disable_extension(EXT_ZCF);
+  } else if (new_f && !prev_f && proc->extension_enabled(EXT_ZCA)) {
+    proc->enable_extension(EXT_ZCF);
+  }
```
Comment (scottj97) [inline riscv/csrs.cc]: I assume Zca requires Zcf? This is not the best place to express that restriction, and I suspect it's a duplicate of the ISA parser which also checks that restriction.

If the ISA object will not allow Zcf without Zca then I don't see the harm in calling `enable_extension` here, since `isa->extension_enabled(EXT_ZCF)` will still be false.

This entire `if` should be one function call: `set_extension_enable(EXT_ZCF, new_f)`

File: riscv/processor.h [hunk 5]
```diff
@@ -209,8 +209,17 @@ class processor_t : public abstract_device_t
...
+      return isa->extension_enabled(ext) && !extension_disable_table[ext];
   }
+
+  void enable_extension(unsigned char ext) {
+    extension_disable_table[ext] = false;
+  }
+
+  void disable_extension(unsigned char ext) {
+    extension_disable_table[ext] = true;
+  }
```
Comment (scottj97) [inline riscv/processor.h]: Please move these below (or above) `extension_enabled_const()` so we can keep that next to its kin `extension_enabled()`.

Also, `extension_enabled_const()` is now subtly broken. See the comment about "assume this can't change" which is no longer a valid assumption.

There needs to be somewhere a list of extensions that are allowed to be dynamically disabled. `disable_extension()` (and the `set_extension_enable()` I suggested) should check that list and `abort()` if any code tries to disable something not in that list. Then `extension_enabled_const()` can check that same list. Only then can assumptions about unchanging extensions be validated, and caught immediately if that should ever change.

File: riscv/processor.h [hunk 6]
```diff
@@ -283,6 +292,8 @@ class processor_t : public abstract_device_t
   bool halt_on_reset;
   std::vector<bool> impl_table;
 
+  std::vector<bool> extension_disable_table;
```
Comment (scottj97) [inline riscv/processor.h]: `vector<bool>` is weird and slow. Simply:
```
bool extension_disable_table[256] = {};
```

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/csrs.cc (hunk 3 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: Yeah. I attempted re-enable them in following code. But something wrong in their logic  I'll fix it later.

File: riscv/csrs.cc (hunk 4 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: Zcf requires Zca. We cannot enable Zcf without Zca enabled. The corner case is  isa=CF, if we disable C and F firstly, and enable F without C later.

File: riscv/processor.h (hunk 5 above)
Comment (liweiwei90) [reply riscv/processor.h]: OK. I'll update it.

File: riscv/processor.h [hunk 7]
```diff
@@ -209,16 +209,24 @@ class processor_t : public abstract_device_t
...
   }
   // Is this extension enabled? and abort if this extension can
   // possibly be disabled dynamically. Useful for documenting
   // assumptions about writable misa bits.
   bool extension_enabled_const(unsigned char ext) const {
     if (ext >= 'A' && ext <= 'Z')
       return state.misa->extension_enabled_const(ext);
-    else
+    else {
+      if (ext == EXT_ZCA || ext == EXT_ZCF || ext == EXT_ZCD) {
```
Comment (scottj97) [inline riscv/processor.h]: This is duplicate code. There must be one definition of this list. Otherwise someone will change one without realizing there's two.

File: riscv/csrs.cc (hunk 4 above)
Comment (scottj97) [reply riscv/csrs.cc]: This is a really confusing block of code. We need a way to express these dependencies succinctly instead of through jumbled logic like this. For example we have `misa_csr_t::dependency()` which enables understandable adjustments of the `misa` value like so:
```
  adjusted_val = dependency(adjusted_val, 'D', 'F');
  adjusted_val = dependency(adjusted_val, 'Q', 'D');
  adjusted_val = dependency(adjusted_val, 'V', 'D');
```

**Review by scottj97 — CHANGES_REQUESTED** [review]

**Comment by scottj97** [comment]
There should not be foxtrot merges like [ffbd5bb](https://github.com/riscv-software-src/riscv-isa-sim/pull/1176/commits/ffbd5bb3d013cd9dd804985043a65213b8a221b3) in PRs. Use rebase if you want to include the latest master.

File: riscv/processor.h (hunk 7 above)
Comment (liweiwei90) [reply riscv/processor.h]: OK. I'll try to update this.

File: riscv/processor.h [hunk 8]
```diff
@@ -205,20 +205,33 @@ class processor_t : public abstract_device_t
...
   // Is this extension enabled? and abort if this extension can
   // possibly be disabled dynamically. Useful for documenting
   // assumptions about writable misa bits.
   bool extension_enabled_const(unsigned char ext) const {
     if (ext >= 'A' && ext <= 'Z')
       return state.misa->extension_enabled_const(ext);
-    else
+    else {
+      assert(!extension_changable(ext));
       return isa->extension_enabled(ext);  // assume this can't change
```
Comment (scottj97) [inline riscv/processor.h]: Remove now-obsolete comment

**Review by scottj97 — APPROVED** [review]
This is much improved.

One little nit to clean up and this is good.

File: riscv/processor.h (hunk 8 above)
Comment (liweiwei90) [reply riscv/processor.h]: OK. I'll remove the "assume ..."

**Comment by scottj97** [comment]
@liweiwei90 please rebase onto the latest master. @aswaterman any reason not to merge this as soon as that's complete?

File: riscv/csrs.cc [hunk 9]
```diff
@@ -620,8 +620,10 @@ reg_t misa_csr_t::dependency(const reg_t val, const char feature, const char dep
 }
 
 bool misa_csr_t::unlogged_write(const reg_t val) noexcept {
+  bool zca_changable = proc->extension_changable(EXT_ZCA);
```
Comment (aswaterman) [inline riscv/csrs.cc]: "changeable" is misspelled here and elsewhere; let's fix that before merging.

File: riscv/processor.h [hunk 10]
```diff
@@ -283,6 +296,8 @@ class processor_t : public abstract_device_t
   bool halt_on_reset;
   std::vector<bool> impl_table;
 
+  bool extension_disable_table[256] = {};
```
Comment (aswaterman) [inline riscv/processor.h]: To match recent changes to `isa_parser_t`, and to support more than 256 extensions, please use the data structure `std::bitset<NUM_ISA_EXTENSIONS>` instead of `bool[256]`.

File: riscv/processor.h [hunk 11]
```diff
@@ -205,20 +205,33 @@ class processor_t : public abstract_device_t
...
+              isa->extension_enabled(EXT_ZCMP) ||
+              isa->extension_enabled(EXT_ZCMT))) ||
+           ext == EXT_ZCF || ext == EXT_ZCD;
+  }
   bool extension_enabled(unsigned char ext) const {
     if (ext >= 'A' && ext <= 'Z')
       return state.misa->extension_enabled(ext);
     else
-      return isa->extension_enabled(ext);
+      return isa->extension_enabled(ext) && !extension_disable_table[ext];
```
Comment (aswaterman) [inline riscv/processor.h]: I would prefer that we check only the data structure `extension_disable_table` here, rather than checking two things.  This is better for performance and IMO is easier to reason about.

The implication would be that `extension_enable_table[ext]` always be false when `isa->extension_enabled(ext)` is false.

File: riscv/processor.h (hunk 10 above)
Comment (aswaterman) [reply riscv/processor.h]: I think it would be easier to understand if we invert the polarity of this table (`extension_enable_table`).  This makes more sense when combined with my comment above https://github.com/riscv-software-src/riscv-isa-sim/pull/1176/files#r1092287066

**Comment by aswaterman** [comment]
@scottj97 I think the premise is fine, but I'd like to re-review after @liweiwei90 has rebased it and has addressed my suggestions.

File: riscv/csrs.cc (hunk 9 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: OK. I'll fix it later.

File: riscv/processor.h (hunk 10 above)
Comment (liweiwei90) [reply riscv/processor.h]: OK. I'll try to update these.

File: riscv/processor.h [hunk 12]
```diff
@@ -234,8 +241,14 @@ class processor_t : public abstract_device_t
...
       return state.misa->extension_enabled_const(ext);
-    else
-      return isa->extension_enabled(ext);  // assume this can't change
+    else {
+      assert(!extension_changeable(ext));
+      return isa->extension_enabled(ext);
+    }
+  }
+  void set_extension_enable(unsigned char ext, bool enable) {
+    assert(enable || extension_changeable(ext));
```
Comment (scottj97) [inline riscv/processor.h]: This doesn't sound right. If the extension is disabled and not changeable, this will incorrectly allow it to be enabled.

File: riscv/processor.h [hunk 13]
```diff
@@ -216,14 +216,21 @@ class processor_t : public abstract_device_t
   bool any_custom_extensions() const {
     return !custom_extensions.empty();
   }
+  bool extension_changeable(unsigned char ext) const {
+    return (ext == EXT_ZCA &&
+            !(isa->extension_enabled(EXT_ZCB) ||
+              isa->extension_enabled(EXT_ZCMP) ||
+              isa->extension_enabled(EXT_ZCMT))) ||
```
Comment (scottj97) [inline riscv/processor.h]: Should these be `extension_enabled_const()`?

I know that method calls `extension_changeable()` so there is a potential for endless recursion, but I think it should be okay as long as `ext != EXT_ZCA` on those recursive calls.

**Comment by scottj97** [comment]
This PR needs to discuss what happened in #1203 because it seems to undo that change.

**Review by scottj97 — CHANGES_REQUESTED** [review]
I know I approved this before, but either I'm paying more attention today, or something has changed. This is now doing quite a bit of refactoring, and also making behavioral changes (misa.C is now writable). This needs to be split up into multiple commits, to separate the refactoring from the behavior changes, and so the commit messages can clearly explain exactly what's changing at each step.

File: riscv/processor.h (hunk 12 above)
Comment (liweiwei90) [reply riscv/processor.h]: Yeah. I didn't take other extensions into consideration. I'll fix this later.

File: riscv/processor.h (hunk 13 above)
Comment (liweiwei90) [reply riscv/processor.h]: This changeable function is only for extensions > 'Z' . So `extension_enabled_const()` will finally call` isa->extension_enabled() ` too.

**Comment by liweiwei90** [comment]
> This PR needs to discuss what happened in #1203 because it seems to undo that change.

Yeah. In previous Zc* implementation, disable C will not disable Zca. So it doesn't works and makes the ma_fetch test fail. So #1203 make 'C' read-only to bypass this question. This patch is also to fix this problem from another way: make 'C' in misa still writable, and conditionally disable Zca when clear C in misa.

**Review by scottj97 — APPROVED** [review]
I think the ISA is quite confusing here, and I have no opinion about the correctness of the ISA or its implementation here, but otherwise this PR ~looks good~.

Edit: except for the two problems I found below!

File: riscv/processor.h (hunk 13 above)
Comment (scottj97) [reply riscv/processor.h]: I was also confused about `proc->extension_enabled()` versus `isa->extension_enabled()`.

File: riscv/csrs.cc [hunk 14]
```diff
@@ -607,6 +607,7 @@ misa_csr_t::misa_csr_t(processor_t* const proc, const reg_t addr, const reg_t ma
                         | (1L << ('F' - 'A'))
                         | (1L << ('D' - 'A'))
                         | (1L << ('Q' - 'A'))
+                        | (1L << ('C' - 'A'))
```
Comment (scottj97) [inline riscv/csrs.cc:610]: The comment on line 604 above needs updating

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/csrs.cc [hunk 15]
```diff
@@ -618,6 +619,11 @@ reg_t misa_csr_t::dependency(const reg_t val, const char feature, const char dep
 }
 
 bool misa_csr_t::unlogged_write(const reg_t val) noexcept {
+  // the write is ignored if increasing IALIGN would misalign the PC
+  if (!(val & (1L << ('C' - 'A'))) && (state->pc & 2) &&
+      proc->extension_changeable(EXT_ZCA))
+    return false;
```
Comment (scottj97) [inline riscv/csrs.cc]: Will this work correctly with `--isa=rv64imafdc`? It seems that `extension_changeable(EXT_ZCA)` will be false in that case, so this would incorrectly allow C to be disabled for that ISA even when the PC is misaligned.

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/csrs.cc (hunk 14 above)
Comment (liweiwei90) [reply riscv/csrs.cc:610]: OK.

File: riscv/csrs.cc (hunk 15 above)
Comment (liweiwei90) [reply riscv/csrs.cc]: `extension_changeable(EXT_ZCA) `will be true in this case. 'C' will imply Zca, so`isa->extension_enabled(EXT_ZCA)` and `isa->extension_enabled('C') `will be true. And `isa->extension_enabled(EXT_ZCB) || isa->extension_enabled(EXT_ZCMP) || isa->extension_enabled(EXT_ZCMT)` will be false in this case.

**Review by scottj97 — APPROVED** [review]

**Review by aswaterman — CHANGES_REQUESTED** [review]
My concern is that we have complicated logic spread across multiple places.  I tried to build on this PR to keep the logic isolated in the misa-write logic, and if I got it correct, it's definitely simpler.  Can y'all take a look, and if it seems correct (or can be made correct), then can you incorporate it, @liweiwei90?

**Comment by aswaterman** [comment]
Here's my attempted simplification: https://github.com/riscv-software-src/riscv-isa-sim/commit/a883a5839dc1f09e5fc8aa80e563f4c841c4a4bc

**Comment by aswaterman** [comment]
@liweiwei90 @scottj97 I believe I've fixed the problems with my proposed version and have pushed to the https://github.com/riscv-software-src/riscv-isa-sim/tree/plctlab-plct-zce-fix2 branch again; see https://github.com/riscv-software-src/riscv-isa-sim/commit/c790cfd3b7e87f69b6c43018f1ccdbf188188d49

This version fixes the other Zc* extensions.  The logic is that they can only be enabled if they are statically enabled, Zca is dynamically enabled, and any other extension they depend upon is dynamically enabled.

This version also keeps the logic in one place, and simplifies usage of processor_t:: set_extension_enable by having its implementation mask the argument with isa_t::extension_enabled.

Despite the annoyance of the specification and the somewhat limited benefit, I'm getting comfortable with this version being simple enough that it isn't too much of a burden to continue supporting.

**Review by aswaterman — APPROVED** [review]

**Comment by aswaterman** [comment]
@scottj97 can you re-review?

File: riscv/processor.h [hunk 16]
```diff
@@ -234,8 +234,12 @@ class processor_t : public abstract_device_t
   bool extension_enabled_const(isa_extension_t ext) const {
     if (ext >= 'A' && ext <= 'Z')
       return state.misa->extension_enabled_const(ext);
-    else
-      return isa->extension_enabled(ext);  // assume this can't change
+    else {
+      return isa->extension_enabled(ext);
+    }
```
Comment (scottj97) [inline riscv/processor.h:239]: The goal of `extension_enabled_const()` (versus plain `extension_enabled()`) is to document and enforce an assumption that an extension will not be dynamically changing during the life of the target program. A future change (such as this one, which allows some extensions to be newly disabled) could invalidate those assumptions, and without `extension_enabled_const()`, could be subtly broken.

When `--isa=rv64imafdc`, following a `csrw misa.C=0` it looks like `extension_enabled_const(EXT_ZCA)` will return `true`, even though this extension is not actually enabled anymore. It should always either fire an assertion or return the same value as `extension_enabled()`.

After this change, since `extension_enabled(EXT_ZCA)` can change during the program, we want `extension_enabled_const(EXT_ZCA)` to fire an assertion -- at least when C is enabled in the ISA (and I think it's okay to even when it's not).

Unfortunately I don't see a robust and easy way to do this without duplicating some of the logic in `misa_csr_t::unlogged_write()` about which extensions can be dynamically disabled.

We could keep track of which extensions have been assumed const by recording that in a new `bitset`, and keep track of which extensions have been dynamically disabled by recording that in another new `bitset`, and then assert in both `extension_enabled_const()` and `set_extension_enable()` by checking those new variables. That would be robust but it seems overly complicated.

**Review by scottj97 — CHANGES_REQUESTED** [review]
I think this should be squashed/rebased into two commits:

1. Enable dynamic disabling of extensions (beyond single-letter ones already supported via `misa`) -- adds `extension_enable_table` and `set_extension_enable()`
2. Enable `misa.C` toggling

Edit: also please update the description of this PR to match the new behavior

File: riscv/processor.h (hunk 16 above)
Comment (scottj97) [reply riscv/processor.h:239]: I've implemented this in [c9448044](https://github.com/scottj97/riscv-isa-sim/commit/c9448044bd96fe6a5666d68f90137fcbceb90a57) and it's not as bad as I thought.

The new assertions fire because in processor.cc we have this line:
```
  if (proc->extension_enabled_const(EXT_ZCMT))
    csrmap[CSR_JVT] = jvt = std::make_shared<jvt_csr_t>(proc, CSR_JVT, 0);
```

Which needs to be addressed first. If Zcmt is disabled (because `misa.C==0`) should the JVT CSR still be accessible? If yes, this should be changed to `extension_enabled()`. If not, then in addition to that change, `jvt_csr_t::verify_permissions()` needs to consider `extension_enabled(EXT_ZCMT)`.

Comment (aswaterman) [reply riscv/processor.h:239]: Nice job with c9448044bd96fe6a5666d68f90137fcbceb90a57: the dynamic approach is actually quite clean.

Re: `Zcmt`, our only choices are to make `misa.C` read-only when Zcmt is present, or make the change you suggested.  It's more in keeping with this PR's approach to do the latter.  I'll open a new PR since I don't have write permissions on this one.

Comment (liweiwei90) [reply riscv/processor.h:239]: OK. I'll close this PR.
